beacon_node/beacon_chain/src/eth1_cache.rs

```diff
@@ -0,0 +1,478 @@
+/// as this structure stores these values from the `BeaconState` at a `Checkpoint`
+#[derive(Clone)]
+pub struct Eth1CacheData {
+    pub eth1_data: Eth1Data,
+    pub eth1_deposit_index: u64,
+}
+
+impl Eth1CacheData {
+    fn fully_imported(&self) -> bool {
+        self.eth1_deposit_index >= self.eth1_data.deposit_count
+    }
+}
+
+/// Implements map from Checkpoint -> Eth1CacheData
+pub struct CheckpointMap {
+    capacity: usize,
+    // There shouldn't be more than a couple of potential checkpoints at the same
+    // epoch, so searching through a vector for the matching Root should be faster
+    // than using another map from Root->Eth1CacheData
+    store: BTreeMap<Epoch, Vec<(Root, Eth1CacheData)>>,
+}
+
+impl Default for CheckpointMap {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Provides a map of `Eth1CacheData` referenced by `Checkpoint`
+///
+/// ## Cache Queuing
+///
+/// The cache keeps a maximum number of (`capacity`) epochs. Because there may be
+/// forks at the epoch boundary, it's possible that there exists more than one
+/// `Checkpoint` for the same `Epoch`. This cache will store all checkpoints for
+/// a given `Epoch`. When adding data for a new `Checkpoint` would cause the number
+/// of `Epoch`s stored to exceed `capacity`, the data for oldest `Epoch` is dropped
+impl CheckpointMap {
+    pub fn new() -> Self {
+        CheckpointMap {
+            capacity: DEFAULT_ETH1_CACHE_SIZE,
+            store: BTreeMap::new(),
+        }
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        CheckpointMap {
+            capacity: cmp::max(1, capacity),
+            store: BTreeMap::new(),
+        }
+    }
+
+    pub fn insert(&mut self, checkpoint: Checkpoint, eth1_cache_data: Eth1CacheData) {
+        self.store
+            .entry(checkpoint.epoch)
+            .or_insert_with(Vec::new)
+            .push((checkpoint.root, eth1_cache_data));
+
+        // faster to reduce size after the fact than do pre-checking to see
+        // if the current data would increase the size of the BTreeMap
+        while self.store.len() > self.capacity {
+            let oldest_stored_epoch = self.store.keys().next().cloned().unwrap();
+            self.store.remove(&oldest_stored_epoch);
+        }
+    }
+
+    pub fn get(&self, checkpoint: &Checkpoint) -> Option<&Eth1CacheData> {
+        match self.store.get(&checkpoint.epoch) {
+            Some(vec) => {
+                for (root, data) in vec {
+                    if *root == checkpoint.root {
+                        return Some(data);
+                    }
+                }
+                None
+            }
+            None => None,
+        }
+    }
+
+    #[cfg(test)]
+    pub fn len(&self) -> usize {
+        self.store.len()
+    }
+}
+
+/// This cache stores `Eth1CacheData` that could potentially be finalized within 4
+/// future epochs.
+pub struct Eth1Cache {
```

Comment:
Given the plethora of eth1-related caches, perhaps we could name this one something like `Eth1FinalizationCache`?

Reply:
done!